### PR TITLE
Add timezone support for .at()

### PR DESCRIPTION
**Pull request by SijmenHuizenga**
Add a timezone parameter to the `.at()` method:
```python
schedule.every().day.at("12:42", "Asia/Jakarta").do(job)
```

During next-job-calculation, the next run is translated from the set timezone into the local timezone so that all jobs are being scheduled in the same timezone.

This implementation of timezone-support is inspired by #471, difference here is that [pytz](https://pypi.org/project/pytz/) is an optional dependency. Pytz is only required when using timezones. This ensures that people can keep using `schedule` without other dependencies.

What do you think? Feedback & Reviews is welcome!

### CODE REVIEW

File: schedule/__init__.py [hunk 1]
```diff
@@ -226,6 +226,9 @@ def __init__(self, interval: int, scheduler: Scheduler = None):
         # optional time at which this job runs
         self.at_time: Optional[datetime.time] = None
 
+        # optional time zone of the self.at_time field. Only relevant when at_time is not None
+        self.at_time_zone = None
```
Comment (rudSarkar) [inline schedule/__init__.py:230]: I wanted to ask about the optional but the timezone is already set as an option. It can be used without timezone example:

```python
 every().day.at("06:30")
```

If need to timezone just use like:

```python
every().day.at("10:30", "America/New_York")
```

**Comment by SijmenHuizenga** [comment]
Configured tox to run all test twice: First without pytz, then with pytz. This makes sure that the library is in perfectly usable without pytz installed. Test-coverage is measured from the tests that do include pytz.

File: docs/timezones.rst [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+This causes the ``next_run`` and ``last_run`` to always be in Pythons local timezone.
+
+Daylight Saving Time
+~~~~~~~~~~~~~~~~~~~~
+When scheduling jobs with relative time (that is when not using ``.at()``), daylight saving time (DST) is **not** taken into account.
+A job that is set to run every 4 hours might execute after 3 realtime hours when DST goes into effect.
+This is because schedule is timezone-unaware for relative times.
+
+However, when using ``.at()``, DST **is** handed correctly: the job will always run at (or close after) the set timestamp.
+A job scheduled during a moment that is skipped, the job will execute after the cock is moved.
```
Comment (chrimaho) [inline docs/timezones.rst:37]: Please fix typo.

```suggestion
A job scheduled during a moment that is skipped, the job will execute after the clock is moved.
```

File: tox.ini [hunk 3]
```diff
@@ -11,7 +11,13 @@ python =
...
+deps =
+    mock
+    pytest
+    pytest-cov
+    mypy
+    types-pytz
+    pytz: pytz
 commands =
     py.test test_schedule.py schedule -v --cov schedule --cov-report term-missing
     python -m mypy -p schedule
```
Comment (chrimaho) [inline tox.ini:23]: Add this part to fix dependency error for `types-pytz` when running `mypy` tests.

```suggestion
    python -m mypy -p schedule --install-types --non-interactive
```

File: test_schedule.py [hunk 4]
```diff
@@ -17,6 +19,10 @@
     IntervalError,
 )
 
+# Set timezone to Europe/Berlin (CEST) to ensure global reproducibility
+os.environ["TZ"] = "CET-1CEST,M3.5.0,M10.5.0"
+time.tzset()
```
Comment (chrimaho) [inline test_schedule.py:24]: Why is this line here? Can we delete it?

**Review by chrimaho — COMMENTED** [review]
Some comments to help improve this PR.

File: test_schedule.py (hunk 4 above)
Comment (SijmenHuizenga) [reply test_schedule.py:24]: This line ensures changing the timezone is updated throughout python lib as described [here](https://docs.python.org/3/library/time.html#time.tzset): 
```
Although in many cases, changing the TZ environment variable may affect the output of functions like [localtime()](https://docs.python.org/3/library/time.html#time.localtime) without calling [tzset()](https://docs.python.org/3/library/time.html#time.tzset), this behavior should not be relied on.
```

File: test_schedule.py [hunk 5]
```diff
@@ -501,6 +507,58 @@ def test_next_run_time_minute_end(self):
...
+            return
+
+        with mock_datetime(2022, 2, 1, 23, 15):
+            # Current Berlin time: feb-1 23:15 (local)
+            # Current India time: feb-2 03:45
+            # Expected to run India time: feb-2 06:30
+            # Next run Berlin time: feb-2 02:00
+            next = every().day.at("06:30", "Asia/Kolkata").do(mock_job).next_run
+            assert next.hour == 2
+            assert next.minute == 0
```
[the remark this answers is not in the document]
Comment (SijmenHuizenga) [reply test_schedule.py:525]: Let's keep `assert` for now, as is used everywhere in this file. If there are advantages (?) we can update it to use `assertEqual` in another pr :)

File: tox.ini (hunk 3 above)
Comment (SijmenHuizenga) [reply tox.ini:23]: What error are you referring to? I can't find it... 

`types-pytz` is already in the `deps` list so it should already be installed, what does `install-types` do extra?

**Comment by SijmenHuizenga** [comment]
Thank you @chrimaho so much for the review! Super helpful comments :rocket:

File: tox.ini (hunk 3 above)
Comment (chrimaho) [reply tox.ini:23]: Hi @SijmenHuizenga, 

When you run:
```sh
python -m mypy -p schedule
```

Then you will receive the following messages:
```text
schedule\__init__.py:489: error: Library stubs not installed for "pytz" (or incompatible with Python 3.7)
schedule\__init__.py:489: note: Hint: "python3 -m pip install types-pytz"
schedule\__init__.py:489: note: (or run "mypy --install-types" to install all missing stub packages)
schedule\__init__.py:489: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
```

You can check the documentation for `--install-types` here:
https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-install-types

However, if you run:
```sh
python -m mypy -p schedule --install-types
```

You will then get the message:
```text
Installing missing stub packages:
~\schedule\.venv\Scripts\python.exe -m pip install types-pytz

Install? [yN]
```

Which is an issue because it's asking for your interaction. Which is not possible when this script is being run on the GitHub Actions level.

So, therefore, we overcome it like this:
```sh
python -m mypy -p schedule --install-types --non-interactive
```

You can check the documentation for `--non-interactive` here:
https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-non-interactive

Basically, the `types-pytz` is required for `mypy` script to do Unit Tests during the GitHub actions level. It is technically _only_ required at this stage. Meaning that it is _not_ required for the normal function of the `schedule` package. Therefore, it is _not_ advisable to add the `types-pytz` to the `requirements.txt` file.

I hope that makes sense?

Comment (SijmenHuizenga) [reply tox.ini:23]: Aaah yess! That makes perfect sense, thanks. So this error doesn't happen in Github actions is because `types-pytz` is installed before `mypy` runs (as part of dependencies).
